src/main/java/org/springframework/integration/kafka/listener/MetadataStoreOffsetManager.java

```diff
@@ -0,0 +1,207 @@
+/*
+ * Copyright 2015 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+package org.springframework.integration.kafka.listener;
+
+import static com.gs.collections.impl.utility.Iterate.groupBy;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import com.gs.collections.api.RichIterable;
+import com.gs.collections.api.block.function.Function;
+import com.gs.collections.api.block.procedure.Procedure2;
+import com.gs.collections.api.map.MutableMap;
+import com.gs.collections.api.multimap.MutableMultimap;
+import com.gs.collections.impl.map.mutable.ConcurrentHashMap;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.integration.kafka.core.BrokerAddress;
+import org.springframework.integration.kafka.core.ConnectionFactory;
+import org.springframework.integration.kafka.core.Partition;
+import org.springframework.integration.kafka.core.Result;
+import org.springframework.integration.metadata.MetadataStore;
+import org.springframework.integration.metadata.SimpleMetadataStore;
+import org.springframework.util.CollectionUtils;
+
+/**
+ * An {@link org.springframework.integration.kafka.listener.OffsetManager} that persists offsets into
+ * a {@link org.springframework.integration.metadata.MetadataStore}
+ *
+ * @author Marius Bogoevici
+ */
+public class MetadataStoreOffsetManager implements OffsetManager {
+
+	private final static Log LOG = LogFactory.getLog(MetadataStoreOffsetManager.class);
+
+	private final GetLeaderFunction getLeaderFunction = new GetLeaderFunction();
+
+	private String consumerId = "spring.kafka";
+
+	private MetadataStore metadataStore = new SimpleMetadataStore();
+
+	private ConnectionFactory connectionFactory;
+
+	private long referenceTimestamp = -2;
+
+	private MutableMap<Partition, Long> offsets = new ConcurrentHashMap<Partition, Long>();
```

Comment:
Actually, I made a mistake in the previous explanation: the right order of priority is: stored offsets, supplied initial offsets, offsets read from partition state, with the state of the `MetadataStoreOffsetManager` being a transient composite of the previous. 

Only updates made via `updateOffset` are guaranteed to be persistent, and they happen only after the message has been processed, i.e. the `MessageListener` has been invoked on the message. Two clients with the same `consumerId` may not operate on the same partition concurrently through the 'KafkaMessageDrivenChannelAdapter` - if there is some asynchronous fanout downstream as a result of a particular `MessageListener` implementation that, for example, sends out messages to be handled asynchronously, the message is considered to be 'processed' once the task has been delegated to the `Executor`.

Therefore, a starting client will try to resume a given partition from where a client with the same consumerId left off. Failing to do that, it will try to either reset to the supplied set of initial offsets, or the state of the partition, in that order. 

(I think that the current implementation does not implement the above fully correctly after some updates this weekend and need to review it - but this is the intent)